internal/store/driver.go

```diff
@@ -642,8 +844,424 @@ func (d *Driver) calculateHTTPSEdges() map[string]ingressv1alpha1.HTTPSEdge {
+		return err
+	}
+
+	if err := d.handleHTTPHeaderFilterAdd(filter.Add, actions, requestRedirectHeaders); err != nil {
+		return err
+	}
+
+	if err := d.handleHTTPHeaderFilterSet(filter, actions, requestRedirectHeaders); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (d *Driver) handleHTTPHeaderFilterRemove(headersToRemove []string, actions *Actions) error {
+	if len(headersToRemove) == 0 {
+		return nil
+	}
+
+	removeHeaders, err := json.Marshal(RemoveHeadersConfig{Headers: headersToRemove})
+	if err != nil {
+		d.log.Error(err, "cannot convert headers to json", "headers", headersToRemove)
+		return err
+	}
+
+	actions.endpointActions = append(actions.endpointActions, ingressv1alpha1.EndpointAction{
+		Type:   "remove-headers",
+		Config: removeHeaders,
+	})
+
+	return nil
+}
+
+func (d *Driver) handleHTTPHeaderFilterAdd(headersToAdd []gatewayv1.HTTPHeader, actions *Actions, requestRedirectHeaders map[string]string) error {
+	if len(headersToAdd) == 0 {
+		return nil
+	}
+
+	config := AddHeadersConfig{Headers: make(map[string]string)}
+	for _, header := range headersToAdd {
+		config.Headers[string(header.Name)] = header.Value
+	}
+
+	if requestRedirectHeaders != nil {
+		for k, v := range config.Headers {
+			requestRedirectHeaders[k] = v
+		}
+	}
+
+	addHeaders, err := json.Marshal(config)
+	if err != nil {
+		d.log.Error(err, "cannot convert headers to json", "headers", headersToAdd)
+		return err
+	}
+
+	actions.endpointActions = append(actions.endpointActions, ingressv1alpha1.EndpointAction{
+		Type:   "add-headers",
+		Config: addHeaders,
+	})
+
+	return nil
+}
+
+func (d *Driver) handleHTTPHeaderFilterSet(filter *gatewayv1.HTTPHeaderFilter, actions *Actions, requestRedirectHeaders map[string]string) error {
+	if filter == nil {
+		return nil
+	}
+	removeHeaders := []string{}
+	for _, header := range filter.Set {
+		removeHeaders = append(removeHeaders, string(header.Name))
+	}
+
+	if err := d.handleHTTPHeaderFilterRemove(removeHeaders, actions); err != nil {
+		return err
+	}
+
+	if err := d.handleHTTPHeaderFilterAdd(filter.Set, actions, requestRedirectHeaders); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+type URLRedirectConfig struct {
+	To         *string `json:"to"`
+	From       *string `json:"from"`
+	StatusCode *int    `json:"status_code"`
+	// convert to response headers
+	Headers map[string]string `json:"headers"`
+}
+
+func (d *Driver) createUrlRedirectConfig(from string, to string, requestHeaders map[string]string, statusCode *int, actions *Actions) error {
+	urlRedirectAction := URLRedirectConfig{
+		To:         &to,
+		From:       &from,
+		StatusCode: statusCode,
+		Headers:    requestHeaders,
+	}
+	config, err := json.Marshal(urlRedirectAction)
+
+	if err != nil {
+		d.log.Error(err, "cannot convert request redirect filter to json", "HTTPRequestRedirectFilter", urlRedirectAction)
+		return err
+	}
+	actions.endpointActions = append(
```

Comment:
be great if we hide this behind `Add` 🤷

Reply:
I'll cut a new ticket to clean up the rest